### PR TITLE
upgrade: Extend the deployment prechecks

### PR DESCRIPTION
**Pull request by jsuchome**
1. Make really sure controller nodes are HA
2. Make sure compute nodes do not have controller roles.

1+2 are the checks that when failing, still allow "normal" upgrade

3. Make sure nova-compute cannot go before nova-controller

3 is a special case of 2, (as described in https://github.com/crowbar/crowbar-core/pull/1109#issuecomment-285008466), but does not allow any kind of upgrade. So it belongs to extra method with `required = true`

### CODE REVIEW

**Comment by jsuchome** [comment]
[S57P5 (3) S56P14 Add the precheck for compute-controller mixed setup](https://trello.com/c/14imamZ1/153-s57p5-3-s56p14-add-the-precheck-for-compute-controller-mixed-setup)

File: crowbar_framework/app/models/api/crowbar.rb [hunk 1]
```diff
@@ -291,6 +291,33 @@ def ha_config_check
...
+              b = r.barclamp
+              next if BarclampCatalog.category(b) != "OpenStack"
+              ro = BarclampCatalog.run_order(b)
+              wrong_roles.push role if ro < nova_order
+            end
+            ret = { controller_roles: { node: node.name, roles: wrong_roles } } if wrong_roles.any?
+          end
+        end
+        ret
+      end
```
Comment (rhafer) [inline crowbar_framework/app/models/api/crowbar.rb]: Any chance of a unit test for this function?

Comment (jsuchome) [inline crowbar_framework/app/models/api/crowbar.rb]: it definitely should be there...

Comment (jsuchome) [inline crowbar_framework/app/models/api/crowbar.rb]: And it's there.

**Pull request merged**